openmc/surface.py

```diff
@@ -94,6 +94,23 @@ def __repr__(self):
 
         return string
 
+    def _key(self):
+        coeffs = list(sorted(self.coefficients.items()))
+        coeff_tup = tuple(j for i in coeffs for j in i)
+        return (self.type, self.boundary_type) + coeff_tup
+
+    def __hash__(self):
+        return hash(self._key())
```

Comment:
This violates an assumption that Python makes about hashable objects, which is that the [hash value is immutable](https://docs.python.org/3/reference/datamodel.html#object.__hash__). Because our surface classes are mutable, that means this key can change, and hence the hash value can change. This can result in very strange behavior, such as the following:
```Python
>>> s1 = openmc.Sphere()
>>> s2 = openmc.Sphere()
>>> collection = {s1, s2}
>>> s1 in collection
True
>>> s1.r = 3
>>> s1 in collection
False
```

Reply:
Ah, bummer. Would an acceptable solution be to remove the __hash__ method and compare __repr__ for equality with the name and ID fields stripped or go back to a _key variable and compare those keys for equality?

Reply:
Yes, something like that would work. Removing `__hash__` means that you won't be able to use a surface object in a set or as the key in a dictionary. Additionally removing `__eq__` means that you would be able to use them in a set/dict, but the hash value is essentially just based on the memory location of the object, which means that two objects that are the same other than their ID will not go to the same dict key. But, you can probably get by just using the ID as the key in a dict.

Reply:
@paulromano I've removed the `__hash__` method for the `Surface` class, but that means I needed to remove `__eq__` and  `__ne__` as well since defining those without defining `__hash__` means the `Surface` can't be used as a key in a set which breaks the memoization when translating to xml elements. Unfortunately, this makes it take a lot longer to find all the duplicates, since it requires multiple iterations through all the surfaces, but maybe I'm missing something that could simplify this.

Reply:
Actually, I could implement `__eq__` as long as `__hash__` is explicitly set to the parent class `__hash__` method - in the case of `Surface` I would assume that would be the `IDManagerMixin` class?